crypto/evp/evp_enc.c

```diff
@@ -930,9 +930,15 @@ int EVP_CIPHER_CTX_set_key_length(EVP_CIPHER_CTX *c, int keylen)
     params[0] = OSSL_PARAM_construct_int(OSSL_CIPHER_PARAM_KEYLEN, &keylen);
     ok = evp_do_ciph_ctx_setparams(c->cipher, c->provctx, params);
 
-    if (ok != EVP_CTRL_RET_UNSUPPORTED)
-        return ok;
+    if (ok == EVP_CTRL_RET_UNSUPPORTED)
+        goto legacy;
 
+    if (ok > 0 && EVP_CIPHER_CTX_key_length(c) != keylen) {
+        EVPerr(0, EVP_R_INVALID_KEY_LENGTH);
```

Comment:
This error seems wrong...  we don't know about the validity of the given key length, we only know that for some reason, the implementation ignored it, i.e. this is similar `ok == EVP_CTRL_RET_UNSUPPORTED`.  If the key length is *actually* invalid, then `ok` would be false (zero).

also, do note that in such a case, the 1.1.1 `EVP_CIPHER_CTX_set_key_length()` returns -1.  Returning 0 like you do below is arguably a breaking change in this case